File: src/test/java/calculator/TestOperation.java
package calculator;

//Import Junit5 libraries for unit testing:
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.*;
import visitor.Counter;

import java.util.Arrays;
import java.util.List;

class TestOperation {

	private Operation o;
	private Operation o2;

	@BeforeEach
	void setUp() throws Exception {
		List<Expression> params1 = Arrays.asList(new MyNumber(3), new MyNumber(4), new MyNumber(5));
		List<Expression> params2 = Arrays.asList(new MyNumber(5), new MyNumber(4));
		List<Expression> params3 = Arrays.asList(new Plus(params1), new Minus(params2), new MyNumber(7));
		o = new Divides(params3);
		o2 = new Divides(params3);
		o.accept(new Counter());
		o2.accept(new Counter());
	}

	@Test
	void testEquals() {
		assertEquals(o,o2);
	}

	@Test
	void testGetDepth() {
		assertEquals(2, o.getDepth());
	}

	@Test
	void testGetOps() {
		assertEquals(3, o.getOps());
	}

	@Test
	void testGetNbs() {
		assertEquals(Integer.valueOf(6), o.getNbs());
	}

}

File: src/test/python/calculator/test_operation.py
import unittest

from src.main.python.calculator.divides import Divides
from src.main.python.calculator.minus import Minus
from src.main.python.calculator.my_number import MyNumber
from src.main.python.calculator.plus import Plus
from src.main.python.visitor.counter import Counter


class TestOperation(unittest.TestCase):

	def setUp(self):
		params1 = [MyNumber(3), MyNumber(4), MyNumber(5)]
		params2 = [MyNumber(5), MyNumber(4)]
		params3 = [Plus(params1), Minus(params2), MyNumber(7)]
		self.o = Divides(params3)
		self.o2 = Divides(params3)
		self.o.accept(Counter())
		self.o2.accept(Counter())

	def test_equals(self):
		self.assertEqual(self.o, self.o2)

	def test_hash_code(self):
		self.assertEqual(hash(self.o), hash(self.o2))

	def test_get_depth(self):
		self.assertEqual(2, self.o.get_depth())

	def test_get_ops(self):
		self.assertEqual(3, self.o.get_ops())

	def test_get_nbs(self):
		self.assertEqual(6, self.o.get_nbs())

if __name__ == '__main__':
	unittest.main()
